### PR TITLE
Fix unfinished step in parallel flow

### PR DESCRIPTION
**Pull request by doontagi**
related issue #3939 

## Motivation
- On https://github.com/spring-projects/spring-batch/issues/3939, when two flows are executed in parallel, I found that even though one flow throw an interruption and terminated the job, the next step of the other flow was executed.
- And this abnormal behavior occurs only when the other flow was added secondly to the job
- Problem is that even if the job containing both flows were terminated and went into STOPPED status, the other flow couldn't recognize the interruption because interruption had been checked by STOPPING status.
https://github.com/spring-projects/spring-batch/blob/9a2b1bbc96f338017aaa5df51214896f5a32d1e5/spring-batch-core/src/main/java/org/springframework/batch/core/repository/support/SimpleJobRepository.java#L290C16-L290C17
Because the job already had been terminated and in STOPPED status, a step in the other flow cannot recognize it and proceeded as like there had been no interruption.
- The reason why this behavior didn't occur when the interruptedFlow added secondly to the job is that secondly added flow was always executed secondly in SplitState. So, if the two step flow would be added firstly, first step in this flow will be processed before the interrupted flow made the job status STOPPED, so it can recognize the interruption when one step has been executed.
- If there are remaining tasks (steps) but the job was terminated due to an exception as like now, there is also a problem that the status of the job may be changed by the remaining tasks.

## Modification
- Changed to wait for all tasks to finish before throwing an exception.
- Changed a job to handle exceptions after waiting for tasks currently being processed to complete. Prevent the job from terminating even though there are remaining tasks.

## Result
- Job, Flow and Steps work same when an exception does not occur.
- Regardless of the order in which flows are added to the Job for parallel processing, if an interruption occurs, only the currently executing step is processed. That is, no further steps are executed after that.
- Close https://github.com/spring-projects/spring-batch/issues/3939

### CODE REVIEW

File: spring-batch-core/src/test/java/org/springframework/batch/core/job/builder/FlowJobBuilderTests.java [hunk 1]
```diff
@@ -419,4 +424,38 @@ public JdbcTransactionManager transactionManager(DataSource dataSource) {
...
+		Flow interruptedFlow = new FlowBuilder<SimpleFlow>("interruptedFlow").start(interruptedStep).build();
+
+		Flow splitFlow = new FlowBuilder<Flow>("splitFlow").split(new SimpleAsyncTaskExecutor())
+			.add(interruptedFlow, twoStepFlow)
+			.build();
+		FlowJobBuilder jobBuilder = new JobBuilder("job", jobRepository).start(splitFlow).build();
+		jobBuilder.preventRestart().build().execute(execution);
+
+		boolean isExecutedNonExecutableStep = countDownLatch.await(1, TimeUnit.SECONDS);
+		assertEquals(BatchStatus.STOPPED, execution.getStatus());
```
Comment (doontagi) [inline spring-batch-core/src/test/java/org/springframework/batch/core/job/builder/FlowJobBuilderTests.java:457]: Finished with `BatchStatus.STOPPED` without executing `nonExecutableStep`.
This test would fail with executing `nonExecutableStep` if it runs on the current branch.

**Comment by doontagi** [comment]
Gentle ping to @fmbenhassine. Could you review this PR? I've fixed https://github.com/spring-projects/spring-batch/issues/3939.
Thanks!

**Comment by fmbenhassine** [comment]
Thank you for your PR!

> Changed to wait for all asynchronous tasks to finish before throwing an exception.

Isn't that not desired when one of the tasks is interrupted? Meaning if one wants to interrupt a step, the idea is not to wait for other steps to finish.

The change in this PR fixes the reported issue, I just want to make sure there is no side effect on the behaviour of step interruption.

**Comment by doontagi** [comment]
Thank you for the comment! And sorry for the late reply. @fmbenhassine 

I've also thought about your concern. And I don't think this change of exception processing location affects the current process of handling interruptions in steps, so has no side effect. 
Because even if an interruption occurs in one of the steps being processed at the same time, the other step may recognize the interruption only after its execution is finished.
https://github.com/spring-projects/spring-batch/blob/425134c74e6b69dde40a81278e4fddbec4778f80/spring-batch-core/src/main/java/org/springframework/batch/core/step/AbstractStep.java#L230
Interruption only can be recognized in catch or final blocks after `doExecute` has been done.

Therefore, even if interrupt occurs in a step, other steps are processed until the end regardless of my modification.

The difference is that the current code immediately propagates the exception generated by interrupt and terminates the job (`STOPPED`) without caring of other steps. Therefore, the remaining flows and steps cannot recognize the interrupt because it recognize it by `STOPPING` status of the job https://github.com/spring-projects/spring-batch/blob/425134c74e6b69dde40a81278e4fddbec4778f80/spring-batch-core/src/main/java/org/springframework/batch/core/repository/support/SimpleJobRepository.java#L287-L294
and continue to be executed. It is a cause of the bug.

In conclusion, there is no change from the view of steps because in the current code, the execution is also not stopped if an exception caused by interrupt occurs in another step. And I think `Job`  shouldn't be stopped if there are running flows and steps.

So I think my changes will prevent other unknown bugs that can occur as flow and step continue to run with job stopped.

Sorry for the lengthy explanation, please comment if you have any other opinions!

**Comment by doontagi** [comment]
Gentle remind @fmbenhassine. Could you check this PR?
Thanks!

**Comment by fmbenhassine** [comment]
Thank you for your feedback! I saw your comment back then and added a thumbs-up emoji and the `status: feedback-provided` tag to acknowledge your update. I should have mentioned the release in which i was planning to include this to avoid you waiting for an update from my side. Anyway, this looks good to me and should go into 5.2.1 planned for tomorrow.

Thank you for going into details with your last update! I love how you explained the step interruption process and how this change should not have a side effect 👍

**Comment by fmbenhassine** [comment]
Rebased and merged as 76789494d7be2f6d658763039046c0cea80a1fe1. Thank you for your contribution!
